### PR TITLE
Adds UUID support to NDT7

### PR DESCRIPTION
**Pull request by pboothe**
Makes ndt7 save its data in a well-formed directory structure.

Also adds UUIDs to the filename and to the metadata.

---
This change is [Reviewable](https://reviewable.io/reviews/m-lab/ndt-server/60)

### CODE REVIEW

File: fdcache/fdcache.go [hunk 1]
```diff
@@ -57,10 +57,14 @@
 package fdcache
 
 import (
+	"fmt"
+	"log"
```
Comment (bassosimone) [inline fdcache/fdcache.go]: Can you use `apex/log` rather than stdlib's `log`?

File: fdcache/fdcache.go [hunk 2]
```diff
@@ -155,3 +159,17 @@ func GetAndForgetFile(conn net.Conn) *os.File {
...
+func GetUUID(conn net.Conn) (string, error) {
+	key := makekey(conn)
+	mutex.Lock()
+	defer mutex.Unlock()
+	entry, found := cache[key]
+	if !found {
+		return "", fmt.Errorf("fd not found")
+	}
+	id, err := uuid.FromFile(entry.Fp)
+	log.Println("UUID:", id)
```
Comment (bassosimone) [inline fdcache/fdcache.go]: This should be `log.Debugf("UUID: %s", id)` where `log` is `apex/log` (or, if it is just meant as a debugging aid, it should be removed before merging).

File: fdcache/fdcache.go [hunk 3]
```diff
@@ -57,10 +57,14 @@
...
 
 import (
+	"fmt"
+	"log"
 	"net"
 	"os"
 	"sync"
 	"time"
+
+	"github.com/m-lab/go/uuid"
```
Comment (bassosimone) [inline fdcache/fdcache.go]: FWIW, I've also reviewed this package and found it reasonable.

File: fdcache/fdcache.go [hunk 4]
```diff
@@ -155,3 +159,17 @@ func GetAndForgetFile(conn net.Conn) *os.File {
...
 }
+
+// GetUUID returns the UUID for a passed-in connection.
+func GetUUID(conn net.Conn) (string, error) {
+	key := makekey(conn)
+	mutex.Lock()
+	defer mutex.Unlock()
+	entry, found := cache[key]
+	if !found {
+		return "", fmt.Errorf("fd not found")
```
Comment (bassosimone) [inline fdcache/fdcache.go]: Could you make an `error` here using `errors.New`? (What is the difference between making an error using `errors.New` and using `fmt.Errorf`?)

File: ndt-server_test.go [hunk 5]
```diff
@@ -112,30 +125,41 @@ func Test_MainIntegrationTest(t *testing.T) {
...
 
+	// Get the datadir
+	dataDir := os.Getenv("DATADIR")
+
 	tests := []struct {
 		name string
 		cmd  string
+		// ignoreData's default value (false) will NOT ignore whether data is
+		// produced. This is good, because it forces tests which ignore their output
+		// data to explicitly specify this fact.
```
Comment (bassosimone) [inline ndt-server_test.go]: Yeah, this is good

File: ndt7/server/results/file.go [hunk 6]
```diff
@@ -16,15 +18,19 @@ type File struct {
...
-	timestamp := time.Now().UTC().Format(format)
-	name := "ndt7-" + timestamp + ".jsonl.gz"
+func newFile(datadir, what, uuid string) (*File, error) {
+	timestamp := time.Now().UTC()
+	dir := path.Join(datadir, what, timestamp.Format("2006/01/02"))
+	err := os.MkdirAll(dir, 0755)
+	if err != nil {
+		return nil, err
+	}
+	name := dir + "/" + timestamp.Format("T15:04:05.000000000Z") + "." + uuid + ".jsonl.gz"
```
Comment (bassosimone) [inline ndt7/server/results/file.go]: Could you keep `ndt7-` in front?

Comment (bassosimone) [reply ndt7/server/results/file.go]: Could you keep the full timestamp in the file name? (IIRC this is how I was told to do for Neubot and is _also_ much more handy if you scrape the files locally in a single directory &mdash; provided you don't have `ext2`)

File: ndt7/server/results/file.go [hunk 7]
```diff
@@ -16,15 +18,19 @@ type File struct {
...
 
 // newFile opens a measurements file in the current working
 // directory on success and returns an error on failure.
-func newFile() (*File, error) {
-	format := "2006-01-02T15:04:05.000000000Z"
-	timestamp := time.Now().UTC().Format(format)
-	name := "ndt7-" + timestamp + ".jsonl.gz"
+func newFile(datadir, what, uuid string) (*File, error) {
+	timestamp := time.Now().UTC()
+	dir := path.Join(datadir, what, timestamp.Format("2006/01/02"))
```
Comment (bassosimone) [inline ndt7/server/results/file.go]: Why do you separate download and upload at the dir level? (I'd have put that in the filename, perhaps)

File: ndt7/server/results/file.go [hunk 8]
```diff
@@ -38,21 +44,27 @@ func newFile() (*File, error) {
...
-	initMetadata(&meta, conn.LocalAddr().String(),
-		conn.RemoteAddr().String(), request.URL.Query(), "download")
-	resultfp, err := newFile()
+	netConn := conn.UnderlyingConn()
+	id, err := fdcache.GetUUID(netConn)
+	if err != nil {
+		return nil, err
+	}
+	initMetadata(&meta, conn.LocalAddr().String(), conn.RemoteAddr().String(), id,
+		request.URL.Query(), "download")
```
Comment (bassosimone) [inline ndt7/server/results/file.go]: If I understand your plan correctly, `s/"download"/what/`

**Comment by pboothe** [comment]
Please take another look.

Also, avoiding the phenomenon of "actually, I no longer understand why this integration test used to pass" is a major motivation behind the development of unit tests to cover "already tested" code.  Integration tests can frequently pass "by accident" when multiple bugs happen to work together just right.

**Comment by pboothe** [comment]
Travis also doesn't support socket cookies (a kernel feature from 2017). So creating files full of data is largely a non-starter because they can't be named because a UUID can't be generated.

File: .travis.yml [hunk 9]
```diff
@@ -2,12 +2,17 @@ language: go
...
 script:
-- go get -v -tags netgo ./cmd/ndt-cloud-client
-- go test -v -coverprofile=ndt.cov -coverpkg=./... -tags netgo ./...
-- $GOPATH/bin/goveralls -coverprofile=ndt.cov -service=travis-ci
+- docker build -f TestDockerfile -t ndt-server-test .
+- docker run -t
+    -e TRAVIS_PULL_REQUEST=$TRAVIS_PULL_REQUEST
+    -e TRAVIS_JOB_ID=$TRAVIS_JOB_ID
+    -e TRAVIS=true
+    -e COVERALLS_SERVICE_NAME=travis-ci
```
Comment (bassosimone) [inline .travis.yml]: IIRC also the `TRAVIS_BRANCH` variable is useful. If coverage is being computed, also `CODECOV_TOKEN` or `COVERALLS_TOKEN` (IIRC) _should_ probably be forwarded to the container. However, perhaps the `COVERALLS_SERVICE_NAME` serves the same purpose here?

File: build.sh [hunk 10]
```diff
@@ -1,6 +1,8 @@
 #!/bin/sh
-# Script to build ndt-server with the correct `go get` flags.
+# Script to build ndt-server with the correct `go get` flags.  This script
+# should be run inside a container.
```
Comment (bassosimone) [inline build.sh]: Why should?

File: test.sh [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+set -ex
+
+# Test the NDT binary
+PATH=${PATH}:${GOPATH}/bin
+if [[ -z ${TRAVIS_PULL_REQUEST} ]]; then
+  go test -v -cover=1 -coverpkg=./... -tags netgo ./...
+else
+  go test -v -coverprofile=ndt.cov -coverpkg=./... -tags netgo ./...
+  /go/bin/goveralls -coverprofile=ndt.cov -service=travis-ci
+fi
```
Comment (bassosimone) [inline test.sh]: Can you explain to me why there are two difference strategies here?

Comment (pboothe) [reply test.sh]: I felt we should only send data to coveralls.io when the tests are run on Travis.

File: .travis.yml (hunk 9 above)
Comment (pboothe) [reply .travis.yml]: I mostly just used an existing example that worked. It looks like coveralls and travis are both happy with the current set of flags?

Comment (bassosimone) [reply .travis.yml]: Definitely.

File: build.sh (hunk 10 above)
Comment (pboothe) [reply build.sh]: Hm. How about: "This script is designed and tested to run as part of the container build process"?

It's not forbidden to just run it, but I don't want to make any promises either.

**Comment by pboothe** [comment]
Updated comments in response to your comments. :)

File: test.sh (hunk 11 above)
Comment (bassosimone) [reply test.sh]: That's reasonable.

**Comment by bassosimone** [comment]
@pboothe :shipit:

**Pull request merged**